mobius-android/src/test/java/com/spotify/mobius/android/ObservableMutableLiveDataTest.java

```diff
@@ -0,0 +1,97 @@
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius.android;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
+import androidx.lifecycle.Lifecycle;
+import com.spotify.mobius.android.MobiusLoopViewModelTestUtilClasses.TestModel;
+import com.spotify.mobius.disposables.Disposable;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ObservableMutableLiveDataTest {
+  @Rule public InstantTaskExecutorRule rule = new InstantTaskExecutorRule();
+
+  private final FakeLifecycleOwner lifecycleOwner1 = new FakeLifecycleOwner();
+
+  private ObservableMutableLiveData<TestModel> underTest = new ObservableMutableLiveData<>();
+
+  @Test
+  public void testDataSendsInactiveStateOnSubscribeAndThenActiveStateWhenObserverGoesActive() {
+    final List<Boolean> receivedEvents = new ArrayList<>(1);
+
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
+    underTest.observe(lifecycleOwner1, model -> {});
+    underTest.subscribe(receivedEvents::add);
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
+
+    assertThat(receivedEvents.size(), equalTo(1));
+    assertThat(receivedEvents.get(0), equalTo(true));
+  }
+
+  @Test
+  public void testDataSendsActiveStateOnSubscribeAndThenInactiveStateWhenObserverGoesInactive() {
+    final List<Boolean> receivedEvents = new ArrayList<>(1);
+
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
+    underTest.observe(lifecycleOwner1, model -> {});
+    underTest.subscribe(receivedEvents::add);
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
+
+    assertThat(receivedEvents.size(), equalTo(1));
+    assertThat(receivedEvents.get(0), equalTo(false));
+  }
+
+  @Test
+  public void testThatDisposeRemovesObserverFromLiveData() {
+    final List<Boolean> receivedEvents = new ArrayList<>(1);
+
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
+    underTest.observe(lifecycleOwner1, model -> {});
+    underTest.subscribe(receivedEvents::add).dispose();
+    lifecycleOwner1.handleLifecycleEvent(Lifecycle.Event.ON_START);
+
+    assertThat(receivedEvents.size(), equalTo(0));
+  }
+
+  private Disposable d = null;
+
+  @Test
+  public void testThatDisposingFromObserverCallbackDoesNotBreak() {
```

Comment:
I don't understand this test case? I would have expected a single `false` to have been added to the `receivedEvents`? Also, it seems like it would be a good idea to verify something like:

1. The subscription is invoked once (leading to a boolean being written to the arraylist)
2. The subscription is then disposed.
3. Additional lifecycle events do not lead to further booleans being written to the arraylist.

Am I misunderstanding something about how the functionality or the test should work?

Reply:
Yeah, I will have a look at this test case again, you have a good point.

Reply:
So I actually wrote this that it doesn't replay the last Lifecycle event as a boolean emission, I believe the reasoning is that this is an internal object that is only created in the `MobiusLoopViewModel` constructor, before there are any lifecycle observers added to it. I think it should be fine, and that the immediate emission creates more problems - but not sure at the moment.

I've added additional check to make sure no other events are recorded into the received list, which also should assure its disposed.